### ISSUE TITLE
Add support for variadic functions (varargs) to GDScript

### ISSUE THREAD

**Issue by MaaaxiKing** [godotengine/godot-proposals#1034]
**Describe the project you are working on:**
Reaction game
**Describe the problem or limitation you are having in your project:**
I can't call a function with a variable amount of arguments (if they don't have a default)!
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I could call a function with a variable amount of arguments. Of course, you should also be able to pass the argument you want, not necessarily in the given order but this is something different: [look here](https://github.com/godotengine/godot-proposals/issues/902)
**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
```
func my_func(*argv):  
	for arg in argv:  
		print (arg) 
    
my_func("Hello", "Welcome", "to", "Godot")  
```
**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No
**Is there a reason why this should be core and not an add-on in the asset library?:**
Yes, it is useful for every project.

*Bugsquad edit (keywords for easier searching): python, args, kwargs*

**Comment by Calinou** [comment]
> **Describe the project you are working on:**

You should fill in that field as well :slightly_smiling_face:

**Comment by MaaaxiKing** [comment]
>Is there a reason why this should be core and not an add-on in the asset library?:

I have no idea for an argument for this but also not against it.

**Comment by itsjavi** [comment]
array pack and unpack for arguments would be also great, similar to Javascript / Node / PHP:

```php

function example(...$items)
{
  foreach($item as $item){
     print_r($item)
  }
}


example(... ["a", "b", "c"]) 
// or
example("a", "b", "c")
```

I think this would be more intuitive and you can name the packed argument however you want

**Comment by MaaaxiKing** [comment]
I don't know if it would be harder to implement **kwargs if this exponentiation operator ** would be added too. What do you think? I hope they won't interrupt each other if both of them will get part of gdscript.

**Comment by shayneoneill** [comment]
Thumps up on this one, "splat" arguments make designing clean APIs much nicer, and allows for better metaprogramming, (Ie, "Get all thse parameters, do something funky with them and then pass them to this function", opening up various functional (ie partial applications, currying, etc) and OOP methodologies (think of how python can pass on *args ,**kwargs to ancestors). Its a supremely useful construct. 

Re: core vs addons. Not sure how you implement language features as plugins lol.

**Comment by me2beats** [comment]
this would be a useful feature
but this would require implementing packing/unpacking arrays first, no?
Btw python uses tuples for that

**Comment by tx350z** [comment]
I've come to the point in my current project where varargs would greatly improve code structure and readability.

I'd prefer the '...' over use of splats to avoid confusion as mentioned in previous comments.

**Comment by Speedphoenix** [comment]
I would like to point out that this would greatly help some builtin components of Godot.

For example the `Tween` node has a method with this signature

```gdscript
bool interpolate_callback(object: Object, duration: float, callback: String, arg1: Variant = null, arg2: Variant = null, arg3: Variant = null, arg4: Variant = null, arg5: Variant = null)
```

With the following description:
> Calls callback of object after duration. arg1-arg5 are arguments to be passed to the callback.

This limits the amount of individual arguments to 5 for the callback, and bloats the signature of `interpolate_callback` (8 arguments is quite a lot in my opinion)

With variadic arguments, the method signature could be reduced to
 ```gdscript
bool interpolate_callback(object: Object, duration: float, callback: String, ...args)
# Or:
bool interpolate_callback(object: Object, duration: float, callback: String, ...args: Variant[])
```
And allow for more than 5 arguments to the callback.

**Comment by Calinou** [comment]
@Speedphoenix That said, Tween is being rewritten to have less methods that take lots of parameters: https://github.com/godotengine/godot/pull/41794

**Comment by Speedphoenix** [comment]
That looks great.

The `tween.tween_callback(callback, params)` will use an array for `params`, so I guess it would mostly be syntactic sugar to make that variadic

**Comment by LightningAA** [comment]
Maybe this is a silly question, but how is:
`my_func(arg0, arg1, arg2)`
Any better than:
`my_func([arg0, arg1, arg2])`
It seems to me like it only saves writing 2 characters in exchange for making GDScript more complicated.

**Comment by pycbouh** [comment]
@LightningAA Well, typing aside the main benefit would be parameter validation. An array is just and array, but multiple distinct parameters, even in a variadic function, can be validated at a signature level. This may be less important if you don't rely on the typing system.

**Comment by LightningAA** [comment]
@pycbouh But what about typed arrays?

**Comment by pycbouh** [comment]
I guess they can help. But as I've said, that's a reason aside from writing code. But coding is also important. If a variadic function has some arguments before the varargs, it may be nicer to write them seamlessly instead of consciously breaking off a set of arguments into an array:
```gdscript
my_func(param0, param1, vararg0, vararg1)
```
vs
```gdscript
my_func(param0, param1, [ vararg0, vararg1 ])
```

**Comment by me2beats** [comment]
> @pycbouh But what about typed arrays that were added to GDScript 2.0?

Typed arrays only work when all items are of one type. Functions arguments usually have different types

**Comment by pycbouh** [comment]
> Typed arrays only work when all items are of one type. Functions arguments usually have different types

Yeah, but varargs would either be of the same type or of a generic type like `Variant`, so it's not that different from a typed array.

**Comment by LightningAA** [comment]
I'm still not completely sold, 
```gdscript
my_func(param0, param1, [vararg0, vararg1])
```
Looks fine to me. I wouldn't really care if varargs were added, especially if they weren't that complicated to implement, but it feels to me like adding unnecessary complexity, "There should be one-- and preferably only one --obvious way to [pass a variable amount of arguments to your function]."

**Comment by pycbouh** [comment]
> "There should be one-- and preferably only one --obvious way to [pass a variable amount of arguments to your function]."

I feel like this is putting that idea to extreme. You can argue that you don't need multiple arguments at all, just pass everything as an array. One way and all.

Variadic functions do not serve the same purpose as passing an array as a parameter. Passing an array is a generic operation, and variadic function tells something very specific to the user about the operation. I don't have a good example handy, I'm afraid, but that's the point of having specific syntax — to pass on additional context. I agree that maybe typed arrays may be a middle ground though.

**Comment by MaaaxiKing** [comment]
Looking at this topic almost exactly a year later, I do not agree anymore with myself concerning the syntax, because I learned Java and got to know how it is made there with the three dots, it does even have much more sense than an awkward star!

**Comment by LightningAA** [comment]
> You can argue that you don't need multiple arguments at all, just pass everything as an array. One way and all.

Yes, but then all your arguments would have to be the same type or any type, and there'd be no assigning names to different parameters, and the amount of parameters would be hard to enforce at compile time. varargs is already pretty much already an array, and all it does is save the user typing 2 characters. 

> Passing an array is a generic operation, and variadic function tells something very specific to the user about the operation.

That's a better argument, but I'd still like to hear some examples 😄

**Comment by matthew-salerno** [comment]
I think the python syntax should be used as that's what gdscript draws most from.
I don't particularly care about *args, but **kwargs would be a lifesaver. For example, let's look at this sklearn class for python:
```sklearn.tree.DecisionTreeRegressor(*, criterion='mse', splitter='best', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features=None, random_state=None, max_leaf_nodes=None, min_impurity_decrease=0.0, min_impurity_split=None, ccp_alpha=0.0)```
Most of the default values are fine, but you may want to change a few in the middle. In Godot, you'd either have to pass a dict and check and fill default values manually in the method, or enter in every argument's default value until you reach the one you want to change when calling. This is quite cumbersome.
That being said, *args could help make things clearer when passing flags:
```
enum {flag1, flag2, flag3}
func myFunc(arg1, arg2, *args):
    if args.has(flag1):
        pass
    elif args.has(flag2):
        pass
    elif args.has(flag3):
        pass
```
it doesn't make a huge difference like kwargs, but it does make the code a little more readable.
IMO, `myfunc(value1, value2, flag3, flag1)` is more recognizable than `myfunc(value1, value2, [flag3, flag1])`, though the latter isn't that bad.

**Comment by LightningAA** [comment]
> I think the python syntax should be used as that's what gdscript draws most from.
> I don't particularly care about *args, but **kwargs would be a lifesaver. For example, let's look at this sklearn class for python:
> 
> ```
> sklearn.tree.DecisionTreeRegressor(*, criterion='mse', splitter='best', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features=None, random_state=None, max_leaf_nodes=None, min_impurity_decrease=0.0, min_impurity_split=None, ccp_alpha=0.0)
> ```
> 
> Most of the default values are fine, but you may want to change a few in the middle. In Godot, you'd either have to pass a dict and check and fill default values manually in the method, or enter in every argument's default value until you reach the one you want to change when calling. This is quite cumbersome.

I'm not a python expert, but that sounds more like named arguements (https://github.com/godotengine/godot-proposals/issues/902), not **kwargs (which I understand to be like *args but it's a dictionary instead of an array).

**Comment by matthew-salerno** [comment]
> > I think the python syntax should be used as that's what gdscript draws most from.
> > I don't particularly care about *args, but **kwargs would be a lifesaver. For example, let's look at this sklearn class for python:
> > ```
> > sklearn.tree.DecisionTreeRegressor(*, criterion='mse', splitter='best', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features=None, random_state=None, max_leaf_nodes=None, min_impurity_decrease=0.0, min_impurity_split=None, ccp_alpha=0.0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Most of the default values are fine, but you may want to change a few in the middle. In Godot, you'd either have to pass a dict and check and fill default values manually in the method, or enter in every argument's default value until you reach the one you want to change when calling. This is quite cumbersome.
> 
> I'm not a python expert, but that sounds more like named arguements (#902), not **kwargs (which I understand to be like *args but it's a dictionary instead of an array).

You are absolutely correct. My mistake. In addition, my point on flags would be better implemented with bit masks.

**Comment by GGalizzi** [comment]
> I'm still not completely sold,
> 
> ```gdscript
> my_func(param0, param1, [vararg0, vararg1])
> ```
> 
> Looks fine to me. I wouldn't really care if varargs were added, especially if they weren't that complicated to implement, but it feels to me like adding unnecessary complexity, "There should be one-- and preferably only one --obvious way to [pass a variable amount of arguments to your function]."

My main problem with the lack of varargs right now, unless I'm missing something in Godot 4. Is that there are internal functions that do use varargs, such as `Callable.call`, I can't use an array like you suggest, to create some function that wraps a `Callable.call` to call some previously assigned Callable with some assigned arguments.

**Comment by arrisar** [comment]
Has there been any news since November? I would love this as a feature. Many people have mentioned before, but there are already many features that use this, so being able to write our own is a no-brainer.

**Comment by Calinou** [comment]
> Has there been any news since November? I would love this as a feature. Many people have mentioned before, but there are already many features that use this, so being able to write our own is a no-brainer.

There are currently no plans to implement this, so it definitely won't happen for 4.0.

**Comment by icecold95** [comment]
```
callable.get_object().callv(callable.get_method(), args)
```
Workaround for `Callable`

**Comment by blipk** [comment]
I'm trying to do something like below, which would be a lot easier if I could override `prints` and call it with `.prints` as well as having variadic arg access so I'm not passing a single array to `.prints`, which causes the output to be encapsulated in string quotation or brackets/braces for other objects. Also, I'm not sure if it changes how the elements of the array are displayed as opposed to passing them as seperate arguments to builtin `prints()`

```
export var DEBUG_LEVEL = 1
func log(args, debug_level = DEBUG_LEVEL, no_repeat = false):
	if debug_level > 0:
		if no_repeat and var2str(args) == var2str(_last_log):
			_last_log = args
			return
		prints(args)
		_last_log = args
```

Language elements like this are useful for reflection which can improve developer usage.

**Comment by blipk** [comment]
> ```
> callable.get_object().callv(callable.get_method(), args)
> ```
> 
> Workaround for `Callable`

This is useful but still you can't create a callable from anything in @GlobalScope or @GDScript
I tried the following but obviously it won't work as @GlobalScope is an annotation and not an identifier:
`var callable = Callable(@GlobalScope, "prints")`

**Comment by tx350z** [comment]
> My proposal is to add a locally scoped `parameters` or alternatively `attributes` or `arguments` variable that is available only inside the function and which contains all the parameters passed to that specific function. This is similar to how JavaScript historically handled this.
> 
> Note that `parameters` in this case does not exist outside the scope of a function. I think this could also be used with a Callable `call()` or a lambda function.

I like the idea of having an "automatic" variable to hold the variable parameters. When trying to described a couple tweaks to the idea I realized it is not clear (indicative) that a method accepts variable parameters. Adding support for the ellipses (...) notation in the method signature clearly indicates the method that accepts a variable number of parameters and should make auto-completion much easier. It would also be nice if the automatic variable containing the varargs has a name that is otherwise not allowed for defined variables (@@varargs, ~~varargs, $$args?) and is easy for the parser to recognize without confusion. This assures there is no accidental collision with developer defined variable names.

**Comment by bsil78** [comment]
In order to be more formal about varargs, maybe an Array in last position could be seen as varargs in a call by GDScript ?

Example : 

func myFun(a:int,others:Array):
  [...]

myFun(1,"b",2,Vector.ZERO) is understood then as myFun(1, ["b",2,Vector.ZERO)

Pëople could argue they want to keep arguments control of analyser doing its job, 
thus some annotation to activate it, like @Varargs, added in front of method could be a solution :

@Varargs
func myFun(a:int,others:Array):
  [...]

**Comment by bsil78** [comment]
> I've come to the point in my current project where varargs would greatly improve code structure and readability.
> 
> I'd prefer the '...' over use of splats to avoid confusion as mentioned in previous comments.

func myFun(a:int, ...others):
[...]

=> would implies others is Array, that seems good too and safer that annotation I proposed, because it is on argument itself

I think it has to be reserved to last parameter to be more simple to parse and check.

**Comment by arrisar** [comment]
> In order to be more formal about varargs, maybe an Array in last position could be seen as varargs in a call by GDScript ?

Making an argument type conditionally collect all remaining args would likely be more prone to causing issues. I'm personally +1 for the spread operator to "collect" the args beyond that point, as is common in other language implementations. 

If the language can imperatively interpret an array type as collecting the remaining args, then it can interpret a spread or other operator to declaratively do it instead.

**Comment by coderbloke** [comment]
Today I would also need this, and found this discussion. I lost a bit between the different suggestion.
For who don't know this in Java, if it helps, they solved vararg this way:
- "..." suffix after type name in method signature, and it is only allowed for the last argument
- inside function body, this parameters are simply an array
- when function called, you can pass an array as parameters, or use "," to separate the values
So as just "..." would be shorthand for array + allowing the caller to use different notation

It would look like this:

```gdscript
func print_sum(of_what: String, values: int...):
    var sum := 0
    for value in values:
        sum += value
    print("%s = %d" % [of_what, sum])
```
So its just an array. Built-in Array supports all Variant types, so can be translated.
But you can use it like this:
```gdscript
var already_spent_money: int = get_it_from_somewhere()
var open_planned_costs: int = get_it_from_elsewhere()
print_sum("foreseen overall cost", already_spent_money, open_planned_costs)
```
But also like this:
```gdscript
var planned_cost_item: Array[int] = get_hundreds_of_int_from_eg_a_database()
print_sum("overall opened plan cost", planned_cost_items)
```
Could work without type hints:
```gdscript
func print_sum(of_what, values...):
    ...
```
And with default values:
```gdscript
func print_sum(of_what: String, values: int... = []):
    ...
```
If "..." does not look nice, with a varargs keyword its the same.
+ As I remember, if zero varargs is given, the function simply gets an empty array (so not null)

**Comment by coderbloke** [comment]
Ahh OK sorry, I see "..." was already in discussion

**Comment by chaeyan** [comment]
this is really important!! please devs consider this...

**Comment by Thebas** [comment]
Any good solution for this example?

`func call_method(method):
	if is_online and is_multiplayer_authority(): 
		rpc(method)
	else:
		call(method)`

was hoping for:

`func call_method(method, varargs : ...Variant):
	if is_online and is_multiplayer_authority(): 
		rpc(method, varargs)
	else:
		call(method, varargs)`

Tried to find a sugared solution but couldn't figure it out. 
Curious.

**Comment by vvvvvvitor** [comment]
Any progress on this? Time and time again I run into the issue of needing varidic functions on my projects, at this point I might just make my own GDExtension to remedy the issue.

**Comment by dalexeev** [comment]
@vvvvvvitor See godotengine/godot#82808. This implements the first part of the proposal, _rest parameter_, which allows you to declare variadic functions (i.e. declare a parameter that packs extra arguments into an array). This part is quite small, I think it is completely ready.

However, many also expect the second part, _spread syntax_, which allows you to unpack arrays into argument/element lists. While this is essentially just syntactic sugar for `callv()` and `array1 + array2`, this is the more complex part. I started working on this, but paused when I ran into some problems with static analysis. I plan to continue working on this in the future.

In any case, this will not be included in 4.3 due to the feature freeze, neither the first nor the second part. As for 4.4, it depends on many factors (will the spread syntax part be ready, will other contributors have time to test and review it).

**Comment by RpxdYTX** [comment]
> Any good solution for this example?
> 
> ```
> func call_method(method):
> 	if is_online and is_multiplayer_authority(): 
> 		rpc(method)
> 	else:
> 		call(method)
> ```
> 
> was hoping for:
> 
> ```
> func call_method(method, varargs : ...Variant):
> 	if is_online and is_multiplayer_authority(): 
> 		rpc(method, varargs)
> 	else:
> 		call(method, varargs)
> ```
> 
> Tried to find a sugared solution but couldn't figure it out. Curious. Maybe this is a good example FOR varargs implementation. Thank you in advance.

At least for now you could pass the additional method parameters as an array and call bindv on method:
```gdscript
func call_method(method, args = []):
    ...
        rpc(method.bindv(args))
    ... 

call_method(print, ["hello"])

or

func call_method(method): ...

call_method(print.bind("hello"))
call_method(print.bindv(["hello"]))
call_method(func(): print("hello"))
```

**Comment by RedMser** [comment]
Great to hear this got merged!

Is there a proposal/PR tracking the spread syntax as well? Or is it currently not considered for development due to the workarounds mentioned?
